usdigitalresponse/usdr-gost
# Vue3 migration

### Ticket #3264 

## Description

This PR migrates Grant Finder and ARPA Reporter from Vue 2 to Vue 3 (with the compatibility build to retain support for most Vue 2 syntax). 

For more details on the decision context and follow-up work, see: 

- [Decision doc and discussion](https://github.com/usdigitalresponse/usdr-gost/pull/3191)
- [Full migration plan](https://docs.google.com/document/d/1V7HMDQzkKuhKvDkRn-4sqlENC4YoMvRIw5BdALWWLuE)

### How to review

The changes in this PR are 90% just following the workflow outlined in the [Vue 3 Migration Guide](https://v3-migration.vuejs.org/migration-build.html#upgrade-workflow). The basic idea is that we switch to using Vue 3, but utilize their built-in Vue 2 compatibility layer (`@vue/compat`), which provides support for the majority of Vue 2 syntax. 

I recommend reviewing commit-by-commit to make things more clear. A quick overview (numbers correspond to the steps in the migration guide, and point to the commit in this PR that implements them): 

1. *[already completed prior to this PR]* Migrated vue-cli to vite (https://github.com/usdigitalresponse/usdr-gost/pull/3095)
2. Update vue package dependencies (f465bebefbb627f5f37738c3ba3866fc9793d732)
3. Update vite.config.js to use Vue 3 with the Vue 2 compatibility build (5af2450839d86781404e86a2d7cb508ce5782ac4)
4. *[skipped; we're not using TypeScript]*
5. Fixed compile-time warnings (fdbd9be36232d46d34d63e8a944d4b7d02cc21c2)
6. *[skipped; fixing runtime warnings pushed to follow-up work]*
7. *[skipped; we don't utilize the `<transition>` component]*
8. Use new global mounting API (182b96b21ac6f0ca2aa908c56c0ad2fa1e5ef55e)
9. Upgraded to `vuex` v4 and updated for [breaking changes](https://vuex.vuejs.org/guide/migrating-to-4-0-from-3-x.html) (8b0433833810313125d5b29337a3ed20d45a07da)
10. Upgraded to `vue-router` v4 and updated for [breaking changes](https://router.vuejs.org/guide/migration/) (732c98fc7123bdfb8e29671fc139e2a82f252a0b)
11. *[planned; burning down the list of deprecation warnings will be follow-up work]*
12. *[planned; finally removing the compatibility build will be follow-up work]*

Other steps taken (beyond what was laid out in the official guide): 

1. Fixed image loading via dynamic `require()` by replacing with `import` (ec9fd17a5b9d8fe08b52a7610f84db04eb556eaf)
2. Fixed router install timing issues, see more detail in the commit message (7dfdb29a82feed61ac11eb54a5d65a8098d7487e)
3. Upgraded bootstrap-vue to version supporting Vue 3 with compat (b1abf9f1001da10088ff63dd3cc89643b331338f)
4. Upgrade to `@vue/test-utils` v2, update testing syntax, and fix/skip failing tests (0ba9a6221cc4b6be29eb0f1fa3777d0a52f9ca05)
   - This commit is a bit bigger than I'd like, but I'm running out of time, so please forgive me. It includes the following: 
   - Upgrading the package.json dependency
   - Updating the test utils syntax (this is most of the diff, and pretty rote — stuff like renaming `propsData` to `props` and moving `stubs` and `mocks` into a `global` config when mounting components for testing)
   - A few tests needed to be skipped because fixing them proved too hairy (these are captured as follow-up tickets and documented in the follow-up work plan)
   - A few tests needed to be massaged to get working in the new setup
5. Replaced `VueBootstrapIcons` plugin (from `bootstrap-vue`) library, which I could not get working on the compatibility build at all, with what seems very close to a drop-in replacement that implements Bootstrap's icon library for Vue 3 (without any reliance on `bootstrap-vue`) (42476c6aa0b301bb9b83564b0b4aa9f3a8b2fafe)
6. Added a helper to send deprecation warnings to datadog so we can capture the follow-up work to solve them, plus ignore the deprecation warnings coming from `bootstrap-vue` components, which would otherwise make up 99% of the console warnings (81da76dea33e7219dd20c48a906cff3940419f82)

## Screenshots / Demo Video

## Testing

### Automated and Unit Tests

### Manual tests for Reviewer

## Checklist
